### PR TITLE
Doxygen for torchbind

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35007 Doxygen for torchbind**

image
image
image
image

Differential Revision: [D20525680](https://our.internmc.facebook.com/intern/diff/D20525680)

### CODE REVIEW

File: torch/custom_class.h [hunk 1]
```diff
@@ -17,52 +17,44 @@
...
-  // Constructor. String argument className_ is the name you would like to
-  // see this class exposed as in Python and TorchScript. For example, if
-  // you pass in "MyStack" here, the class will appear as
-  // `torch.classes.MyStack` in both Python and TorchScript.
+  /// This constructor actually registers the class type.
+  /// String argument `className_` is the name you would like to
+  /// see this class exposed as in Python and TorchScript. For example, if
+  /// you pass in "MyStack" here, the class will appear as
+  /// `torch.classes.MyStack` in both Python and TorchScript.
   class_(std::string className_) : className(std::move(className_)) {
```
Comment (dzhulgakov) [inline torch/custom_class.h]: stupid question - why is it className_ (with underscore)? :)

should the constructor be `explicit`?

File: torch/custom_class.h [hunk 2]
```diff
@@ -96,46 +88,47 @@ class class_ {
...
+  /// Python or TorchScript. This protocol is equivalent to the Pickle
+  /// concept of `__getstate__` and `__setstate__` from Python
+  /// (https://docs.python.org/2/library/pickle.html#object.__getstate__)
+  ///
+  /// Currently, both the `get_state` and `set_state` callables must be
+  /// C++ lambda expressions. They should have the following signatures,
+  /// where `CurClass` is the class you're registering and `T` is some object
+  /// that encapsulates the state of the object.
+  ///
+  ///     __getstate__(intrusive_ptr<CurClass>) -> T
```
Comment (dzhulgakov) [inline torch/custom_class.h]: maybe add example of actual registration (i.e. def_pickle call)?

**Review by dzhulgakov — APPROVED** [review]

File: torch/custom_class.h (hunk 1 above)
Comment (jamesr66a) [reply torch/custom_class.h]: 1) Dunno, horace wrote it like that. I can change it
2) yeah probably

**Review by jlin27 — APPROVED** [review]
The doxygen build changes LGTM

**Pull request closed without merging**